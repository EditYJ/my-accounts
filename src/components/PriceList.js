/* eslint-disable jsx-a11y/anchor-is-valid */
import PropTypes from "prop-types";
import React from "react";
import Ionicon from "react-ionicons";
import { TYPE_INCOME } from "../utility";

const PriceList = ({ items, onModifyItem, onDeleteItem }) => {
  return (
    <ul className="list-group list-grop-flush">
      {items.map(item => (
        <li
          className="list-group-item d-flex justify-content-between align-items-center"
          key={item.id}
        >
          <span className="col-1">
            <Ionicon
              className="rounded-circle show-category"
              fontSize="30px"
              style={{ backgroundColor: "#007bff", padding: "5px" }}
              color={"#fff"}
              icon={item.category.iconName}
            />
          </span>
          <span className="col-5 pl-title" style={{ textAlign: "center" }}>
            {item.title}
          </span>
          <span className="col-2 font-weight-bold pl-price">
            {item.category.type === TYPE_INCOME ? "+" : "-"}
            {item.price}å…ƒ
          </span>
          <span className="col-2 pl-date">{item.date}</span>
          <a
            href="#"
            className="col-1"
            onClick={() => {
              onModifyItem(item);
            }}
          >
            <Ionicon
              className="rounded-circle"
              fontSize="30px"
              style={{ backgroundColor: "#28a745", padding: "5px" }}
              color={"#fff"}
              icon="ios-create-outline"
            />
          </a>
          <a
            href="#"
            className="col-1"
            onClick={() => {
              onDeleteItem(item);
            }}
          >
            <Ionicon
              className="rounded-circle"
              fontSize="30px"
              style={{ backgroundColor: "#dc3545", padding: "5px" }}
              color={"#fff"}
              icon="ios-close"
            />
          </a>
        </li>
      ))}
    </ul>
  );
};

PriceList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired
};
PriceList.defaultProps = {
  onModifyItem: () => {},
  onDeleteItem: () => {}
};

export default PriceList;
