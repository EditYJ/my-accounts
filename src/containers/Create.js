import React, { Component } from 'react'
import CategorySelect from '../components/CategorySelect'

const categories = [
  {
    id: '1',
    name: '旅行',
    type: 'income',
    iconName: 'ios-plane'
  },
  {
    id: '2',
    name: '学习',
    type: 'income',
    iconName: 'ios-plane'
  },
  {
    id: '3',
    name: '理财',
    type: 'income',
    iconName: 'ios-plane'
  },
  {
    id: '4',
    name: '理财',
    type: 'income',
    iconName: 'ios-plane'
  }
]
class Create extends Component {
  constructor(props) {
    super(props)
    this.state = {}
  }
  render() {
    return (
      <React.Fragment>
        <CategorySelect
          categories={categories}
          onSelectCategory={category => {
            console.log('category', category, this.props.match)
          }}
          selectCategory={categories[0]}
        ></CategorySelect>
      </React.Fragment>
    )
  }
}

export default Create
