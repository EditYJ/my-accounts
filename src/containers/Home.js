import React from "react";

import CreateButton from "../components/CreateButton";
import MonthPicker from "../components/MonthPicker";
import PriceList from "../components/PriceList";
import TotalPrice from "../components/TotalPrice";
import ViewTab from "../components/ViewTab";
import {
  LIST_VIEW,
  parseToYearAndMounth,
  CHART_VIEW,
  addZero
} from "../utility";

export const categories = {
  1: {
    id: 1,
    name: "旅行",
    type: "outcome",
    iconName: "ios-plane"
  },
  2: {
    id: 2,
    name: "食物",
    type: "outcome",
    iconName: "ios-pizza"
  },
  3: {
    id: 3,
    name: "理财",
    type: "income",
    iconName: "logo-yen"
  }
};
export const items = [
  {
    id: 1,
    title: "去日本旅游",
    price: 200,
    date: "2019-06-24",
    cid: 1
  },
  {
    id: 2,
    title: "买了一些吃的",
    price: 150,
    date: "2019-06-11",
    cid: 2
  },
  {
    id: 3,
    title: "理财收入",
    price: 200,
    date: "2019-09-10",
    cid: 3
  }
];
const newItem = {
  id: 4,
  title: "新添加的项目",
  price: 300,
  date: "2018-10-10",
  cid: 1
};

class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      items,
      currentDate: parseToYearAndMounth('2019/06/11'),
      tabView: LIST_VIEW
    };
  }

  changeView = view => {
    this.setState({
      tabView: view
    });
  };

  changeDate = (year, month) => {
    this.setState({
      currentDate: {
        year,
        month
      }
    });
  };

  modifyItem = modifiedItem => {
    const modifiedItems = this.state.items.map(item => {
      if (item.id === modifiedItem.id) {
        return { ...item, title: "修改后的标题" };
      } else {
        return item;
      }
    });
    this.setState({
      items: modifiedItems
    });
  };

  createItem = () => {
    this.setState({
      items: [newItem, ...this.state.items]
    });
  };

  deleteItem = deletedItem => {
    const filteredItems = this.state.items.filter(
      item => item.id !== deletedItem.id
    );
    this.setState({
      items: filteredItems
    });
  };

  render() {
    const { items, currentDate, tabView } = this.state;
    const itemsWithCategory = items
      .map(item => {
        item.category = categories[item.cid];
        return item;
      })
      .filter(item => {
        return (
          item.date.includes(
            `${this.state.currentDate.year}-${addZero(
              this.state.currentDate.month
            )}`
          )
        );
      });
    return (
      <React.Fragment>
        <header className="App-header">
          <h2 style={{ textAlign: "center" }}>记账本</h2>
          <div className="row">
            <div className="col">
              <MonthPicker
                year={currentDate.year}
                month={currentDate.month}
                onChange={this.changeDate}
              />
            </div>
            <div className="col d-flex justify-content-center align-items-center">
              <TotalPrice income={200} outcome={520} />
            </div>
          </div>
        </header>
        <div className="content-area py-2 px-2">
          <CreateButton onClick={this.createItem} />
          <ViewTab activTab={tabView} onTabChange={this.changeView} />
          {tabView === LIST_VIEW && (
            <PriceList
              items={itemsWithCategory}
              onModifyItem={this.modifyItem}
              onDeleteItem={this.deleteItem}
            />
          )}
          {tabView === CHART_VIEW && (
            <div>
              <h1>这里是图表</h1>
            </div>
          )}
        </div>
      </React.Fragment>
    );
  }
}

export default Home;
