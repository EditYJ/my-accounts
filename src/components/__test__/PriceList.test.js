import React from "react";
import { shallow } from "enzyme";
import PriceList from "../PriceList";
import { items, categories } from "../../containers/Home";

const itemsWithCategory = items.map(item => {
  item.category = categories[item.cid];
  return item;
});

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn()
};

let wrapper;
describe("test PriceList component", () => {
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props} />);
  });

  it("should render the component to match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct price items length", () => {
    expect(wrapper.find(".list-group-item").length).toEqual(
      itemsWithCategory.length
    );
  });

  it("should render correct icon and price for each item", () => {
    const iconList = wrapper.find(".show-category");
    expect(iconList.length).toEqual(3);
    expect(iconList.map(item => item.props().icon)).toEqual(
      itemsWithCategory.map(item => item.category.iconName)
    );
    expect(wrapper.find(".pl-title").map(item => item.text())).toEqual(
      itemsWithCategory.map(item => item.title)
    );
    expect(wrapper.find(".pl-date").map(item => item.text())).toEqual(
      itemsWithCategory.map(item => item.date)
    );
    expect(wrapper.find(".pl-price").map(item => item.text())).toEqual(
      itemsWithCategory.map(
        item => `${item.category.type === "income" ? "+" : "-"}${item.price}å…ƒ`
      )
    );
  });

  it("should trigger the function callbacks", () => {
    const firstItem = wrapper.find('.list-group-item').first()
    firstItem.find('a').first().simulate('click')
    expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategory[0])
    firstItem.find('a').last().simulate('click')
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0])
  });
});
