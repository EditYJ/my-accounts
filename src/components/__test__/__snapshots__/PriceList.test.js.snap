// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshot 1`] = `
<ul
  className="list-group list-grop-flush"
>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="1"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle show-category"
        color="#fff"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5 pl-title"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      去日本旅游
    </span>
    <span
      className="col-2 font-weight-bold pl-price"
    >
      -
      200
      元
    </span>
    <span
      className="col-2 pl-date"
    >
      2019-06-24
    </span>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="2"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle show-category"
        color="#fff"
        fontSize="30px"
        icon="ios-pizza"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5 pl-title"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      买了一些吃的
    </span>
    <span
      className="col-2 font-weight-bold pl-price"
    >
      -
      150
      元
    </span>
    <span
      className="col-2 pl-date"
    >
      2019-06-11
    </span>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between align-items-center"
    key="3"
  >
    <span
      className="col-1"
    >
      <Ionicon
        beat={false}
        className="rounded-circle show-category"
        color="#fff"
        fontSize="30px"
        icon="logo-yen"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5 pl-title"
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      理财收入
    </span>
    <span
      className="col-2 font-weight-bold pl-price"
    >
      +
      200
      元
    </span>
    <span
      className="col-2 pl-date"
    >
      2019-09-10
    </span>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#28a745",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#fff"
        fontSize="30px"
        icon="ios-close"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#dc3545",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
